### PR TITLE
Cookie-based banner preferences

### PR DESCRIPTION
**Pull request by jimchamp**
Modifies our dismissible banners in the following ways:
1. "Hide banner" preferences are stored in cookies, not `User` preferences
2. Multiple banners can be dismissed at once (useful during periods when both a site banner and My Books page banner co-exist).
3. Logic to hide a banner has been moved to the banner template itself.

### Technical

Client-side code has be reorganized, and some functions have been renamed.  `id`s have been removed from the banner template in order to support multiple banners per page.

### Testing

1. Set an announcement and cookie name in `site/body.html`
2. While logged out, visit any page and ensure that the banner is present.  It should not have the "X".
3. Log into the site.  Ensure that the banner is present, and that it has an "X".
4. Click the "X".  Ensure that the banner is immediately hidden.
5. Refresh the page.  Ensure that the banner is still hidden.
6. Check your cookies, ensuring that the name is as expected.

### CODE REVIEW

File: openlibrary/plugins/openlibrary/api.py [hunk 1]
```diff
@@ -658,13 +658,10 @@ class hide_banner(delegate.page):
...
-        user = accounts.get_current_user()
-
-        user.save_preferences({'hidden-banner': data['storage-key']})
 
-        def response(msg, status="success"):
-            return delegate.RawText(
-                json.dumps({status: msg}), content_type="application/json"
-            )
+        # Set truthy cookie that expires in 30 days:
+        web.setcookie(data['cookie-name'], '1', expires=(3600 * 24 * 30))
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/api.py:663]: I wonder if we want this date to be customizable based on the campaign, e.g. `duration_days=30` by default

Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/api.py:663]:
```suggestion
        DAY_SECONDS = 60 * 60 * 24
        web.setcookie(data['cookie-name'], '1', expires=data.get('cookie-duration-days', 30) * DAY_SECONDS))
```

File: openlibrary/plugins/openlibrary/js/banner/index.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+/**
+ * Add click listeners to all banner dismiss buttons.
+ *
+ * @param {NodeList<HTMLElement>} banners
+ */
+export function initDismissibleBanners(banners) {
+    for (const banner of banners) {
+        const cookieName = banner.dataset.cookieName
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/banner/index.js:32]:
```suggestion
        const cookieName = banner.dataset.cookieName
        const cookieDurationDays = banner.dataset.cookieDurationDays
```

File: openlibrary/plugins/openlibrary/js/banner/index.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+        const dismissButton = banner.querySelector('.page-banner--dismissable-close')
+        dismissButton.addEventListener('click', () => {
+            const successCallback = () => {
+                banner.remove()
+            }
+            setBannerCookie(cookieName, successCallback)
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/banner/index.js:39]:
```suggestion
            setBannerCookie(cookieName, cookieDurationDays, successCallback)
```

File: openlibrary/templates/site/banner.html [hunk 4]
```diff
@@ -1,11 +1,17 @@
-$def with (announcement, storage_key)
+$def with (announcement, cookie_name)
```
Comment (mekarpeles) [inline openlibrary/templates/site/banner.html:1]:
```suggestion
$def with (announcement, cookie_name, cookie_duration_days=30)
```

File: openlibrary/templates/site/banner.html [hunk 5]
```diff
@@ -1,11 +1,17 @@
...
 
-$# Renders a dismissable banner element containing the given announcement
+$# Renders a dismissible banner element containing the given announcement.
+$# If the patron has dismissed the banner in the past, the banner will not
+$# be rendered.
+$#
 $# announcement (str) -- Rendered content that will be displayed in the banner
-$# storage_key (str) -- identifier for the banner. If the user dismisses the banner, storage_key will be added to the user's preferences to prevent the banner from showing again.
+$# cookie_name (str) -- identifier for the banner. If the user dismisses the banner, a cookie with this name is set
```
Comment (mekarpeles) [inline openlibrary/templates/site/banner.html:8]:
```suggestion
$# cookie_name (str) -- identifier for the banner. If the user dismisses the banner, a cookie with this name is set
$# cookie_duration_days (int) -- TTL before cookie expires in days
```

File: openlibrary/templates/site/banner.html [hunk 6]
```diff
@@ -1,11 +1,17 @@
...
 
-<div data-storage-key="$storage_key" id="announcement-banner" class="page-banner--dismissable">
-    <span class="page-banner--dismissable-content">$:announcement</span>
-    $if ctx.user:
-        <button class="page-banner--dismissable-close" id="close-banner">X</button>
-</div>
+$ hide_banner = ctx.user and ctx.user.has_cookie(cookie_name)
+
+$if not hide_banner:
+    <div data-cookie-name="$cookie_name" class="page-banner--dismissable">
```
Comment (mekarpeles) [inline openlibrary/templates/site/banner.html:13]:
```suggestion
    <div data-cookie-name="$cookie_name" data-cookie-duration-days="$cookie_duration_days" class="page-banner--dismissable">
```

File: openlibrary/templates/site/body.html [hunk 7]
```diff
@@ -32,11 +32,9 @@
...
-      $# Using the current date in YYYY-MM-DD format for storage_key will prevent the use of duplicate keys.
-      $ storage_key = ''
-      $if announcement and storage_key:
-        $ hide_banner = ctx.user and ctx.user.preferences().get('hidden-banner') == storage_key
-        $if not hide_banner:
-          $:render_template('site/banner', announcement, storage_key)
+      $# Include date in cookie name to avoid name collisions (something like '{prefix}{YY}{MM}{DD}' would work well)
+      $ cookie_name = ''
+      $if announcement and cookie_name:
+        $:render_template('site/banner', announcement, cookie_name)
```
Comment (mekarpeles) [inline openlibrary/templates/site/body.html:38]:
```suggestion
      $if announcement:
        $:render_template('site/banner', announcement, cookie_name="b2308", cookie_duration_days=30)
```

File: openlibrary/plugins/openlibrary/js/banner/index.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
+/**
+ * Makes API call that sets a "hide banner" cookie with the given name.
+ *
+ * Executes the success callback on successful response.
+ *
+ * @param {string} cookieName
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/banner/index.js:6]:
```suggestion
 * @param {string} cookieName
 * @param {int} cookieDurationDays
```

File: openlibrary/plugins/openlibrary/js/banner/index.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+ * @param {Function} successCallback
+ */
+function setBannerCookie(cookieName, successCallback) {
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/banner/index.js:9]:
```suggestion
function setBannerCookie(cookieName, cookieDurationDays, successCallback) {
```

File: openlibrary/plugins/openlibrary/js/banner/index.js (hunk 3 above)
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/js/banner/index.js:39]:
```suggestion
            setBannerCookie(cookieName, cookieDurationDays, successCallback)
```

File: openlibrary/plugins/openlibrary/api.py [hunk 10]
```diff
@@ -658,13 +658,11 @@ class hide_banner(delegate.page):
...
-
-        user.save_preferences({'hidden-banner': data['storage-key']})
 
-        def response(msg, status="success"):
-            return delegate.RawText(
-                json.dumps({status: msg}), content_type="application/json"
-            )
+        # Set truthy cookie that expires in 30 days:
+        DAY_SECONDS = 60 * 60 * 24
+        web.setcookie(data['cookie-name'], '1', expires=data.get('cookie-duration-days', 30) * DAY_SECONDS))
```
[the remark this answers is not in the document]
Comment (jimchamp) [reply openlibrary/plugins/openlibrary/api.py]: There's an unmatched right paren.  Try this:
`web.setcookie(data['cookie-name'], '1', expires=(data.get('cookie-duration-days', 30) * DAY_SECONDS))`

File: openlibrary/templates/site/body.html [hunk 11]
```diff
@@ -32,11 +32,8 @@
...
       $ announcement = ''
-      $# Using the current date in YYYY-MM-DD format for storage_key will prevent the use of duplicate keys.
-      $ storage_key = ''
-      $if announcement and storage_key:
-        $ hide_banner = ctx.user and ctx.user.preferences().get('hidden-banner') == storage_key
-        $if not hide_banner:
-          $:render_template('site/banner', announcement, storage_key)
+      $# Include date in cookie name to avoid name collisions (something like '{prefix}{YY}{MM}{DD}' would work well)
+      $if announcement:
+        $:render_template('site/banner', announcement, cookie_name="b2308", cookie_duration_days=30)
```
Comment (jimchamp) [inline openlibrary/templates/site/body.html:37]: Strongly advise against setting a default cookie name here.

File: openlibrary/plugins/openlibrary/api.py [hunk 12]
```diff
@@ -658,13 +658,12 @@ class hide_banner(delegate.page):
...
-        user.save_preferences({'hidden-banner': data['storage-key']})
 
-        def response(msg, status="success"):
-            return delegate.RawText(
-                json.dumps({status: msg}), content_type="application/json"
-            )
+        # Set truthy cookie that expires in 30 days:
+        DAY_SECONDS = 60 * 60 * 24
+        cookie_duration_days = data.get('cookie-duration-days', 30)
+        web.setcookie(data['cookie-name'], '1', expires=cookie_duration_days * DAY_SECONDS))
```
Comment (jimchamp) [inline openlibrary/plugins/openlibrary/api.py:665]:
```suggestion
        web.setcookie(data['cookie-name'], '1', expires=(cookie_duration_days * DAY_SECONDS))
```
Balance parens.

File: openlibrary/templates/site/body.html (hunk 11 above)
Comment (mekarpeles) [inline openlibrary/templates/site/body.html:37]:
```suggestion
      $ cookie_name = ''
      $if announcement and cookie_name:
        $:render_template('site/banner', announcement, cookie_name, cookie_duration_days=30)
```

**Review by mekarpeles — APPROVED** [review]
changes lgtm, just need to reploy to testing to test

File: openlibrary/plugins/openlibrary/api.py [hunk 13]
```diff
@@ -658,13 +658,14 @@ class hide_banner(delegate.page):
...
         data = json.loads(web.data())
-        user = accounts.get_current_user()
-
-        user.save_preferences({'hidden-banner': data['storage-key']})
 
-        def response(msg, status="success"):
-            return delegate.RawText(
-                json.dumps({status: msg}), content_type="application/json"
-            )
+        # Set truthy cookie that expires in 30 days:
```
Comment (cclauss) [inline openlibrary/plugins/openlibrary/api.py:662]:
```suggestion
        # Set truthy cookie that expires in 24 hours:
```

**Review by mekarpeles — APPROVED** [review]
lgtm, tested on testing

**Pull request merged**